### ISSUE TITLE
Create a test DB

### ISSUE THREAD

**Issue by mgaonach** [ispyb/py-ispyb#123]
Create a test DB:

- that is light enough to be pushed
- set up locally for tests (as easily as possible)
- set up automatically for CI tests

**Comment by stufisher** [comment]
I suggest using https://gitlab.esrf.fr/ui/daiquiri-docker-testdb and making an equivalent for the pyispyb database structure.

**Comment by stufisher** [comment]
I made a start on this here https://github.com/ispyb/py-ispyb/pull/139

It will now start up a docker container with an empty pydb (https://gitlab.esrf.fr/ui/ispyb-docker-pydb), minimal example data can be added to this and used to test against. What do you think @mgaonach ?

Heres the data file that can be appended to:
https://gitlab.esrf.fr/ui/ispyb-docker-pydb/-/blob/main/sql/data.sql

And heres the minimal data for daiquiri, will need some more data for general ispyb tests:
https://gitlab.esrf.fr/ui/daiquiri/-/blob/master/tests/metadata/db/data.sql

**Comment by mgaonach** [comment]
@stufisher created new repo https://github.com/ispyb/ispyb-docker-pydb

**Comment by clemenbor** [comment]
I was able to push to docker hub using a a personal docker account. Example here:
https://github.com/clemenbor/ispyb-docker-pydb/blob/main/.github/workflows/main.yml

The github action is triggered when a new release/tag is created (I guess the trigger can be changed if needed). I took the example from here:
https://docs.github.com/en/actions/publishing-packages/publishing-docker-images

We might need a general "ispyb" docker hub account for the collaboration so all the images are published using that account.

**Comment by mgaonach** [comment]
From meeting:
- Proceed with this testing solution in the short term to get CI up and running
- Need to also think about a second test solution for bigger scale tests (#156)
- Interest from GP to use the database outside of automatic tests as a mock-up database to get ISPyB running without having a real database available. -> when available, mention it in the getting started steps as a demo database.

**Comment by antolinos** [comment]
Just for completion because I explained very badly myself on Yesterday. My comment was that the initial state of the database depends on these inserts:
https://gitlab.esrf.fr/ui/daiquiri/-/blob/master/tests/metadata/db/data.sql

However, nobody warranties that this initial state of the database is a valid state. For instance: these inserts might  populated columns and.or tables that will never be populated by ISPyB.

In order to enforce that the tests are based in a valid state of the database we should use the standard ingestion procedures. e.g: endpoint that synch with the user portal to retrieve proposal, sessions, etc.. and the endpoint for registering experimental data.

**Comment by mgaonach** [comment]
@antolinos 
How about using these procedures once to fill a database, then make a dump and use the result to replace this `data.sql`? 

In my opinion, all tests should be independent of each other. For instance, if we need to run the ingestion tests before we can run the reading test, then on failure we can't know which side the problem comes from without having to dive into the code. A single independent test allows you to know immediately and precisely what failed.
Also sometimes when developing you might want to run a single test without having to run the whole scenario.

What do you think?

**Comment by antolinos** [comment]
I am not discussing the isolation of the tests. You can always drop and populate the database before and after each test to enforce the isolation. The question is how data is populated.
I am just pointing that the testing suite ~can not be based~ would be more realistic if there are not based in a set of arbitrary inserts that might or might not represent a valid data state.  

> How about using these procedures once to fill a database, then make a dump and use the result to replace this data.sql?

Indeed, in my opinion that would be more correct.

**Comment by stufisher** [comment]
The database state is valid because we define what the inserts are. They are not random, but populated to test specific features in the api, i created each insert manually to test a specific part of the db. These are the absolute basics to test certain features. 

They should be expanded on once we want to test more advanced features, there is no problem to add inserts that come from the acquisition client as well

**Comment by antolinos** [comment]
I know that they are not random but also that we can make mistakes. Currently, It is rather simple but at medium/long term once we tests all techniques, pipelines and specificity of each site then it will be much difficult or impossible to verify that the state is valid.
In other words, there are now 32 inserts, will you be sure that the state is valid when the file contains 300, 600 or 1000 inserts?
These data are the foundations of the testing suite therefore it is critical.

**Comment by mgaonach** [comment]
I think using the ingestion procedures to fill a DB and then using a dump to build the docker image is the best compromise.
This way, we can be sure that every test starts with a set of valid data, independently of what is working and what might be broken.
Also, another advantage I see of this vs filling/dropping the DB with the procedures at every test is the execution speed. Dropping and building the database takes time and with this it's already done.
